src/v/cloud_topics/level_zero/gc/level_zero_gc.cc

```diff
@@ -77,19 +77,183 @@ class epoch_source_cluster_impl : public level_zero_gc::epoch_source {
+             */
+            co_await seastar::maybe_yield();
+
+            if (as && as->abort_requested()) {
+                co_return std::unexpected("Abort requested");
+            }
+        }
+
+        co_return result;
+    }
+
     seastar::future<std::expected<std::optional<cluster_epoch>, std::string>>
-    max_gc_eligible_epoch(seastar::abort_source*) override {
+    max_gc_eligible_epoch(seastar::abort_source* as) override {
+        /*
+         * First retrieve a consistent snapshot of cloud topic partitions. This
+         * establishes a set of partitions from which we must obtain an epoch
+         * bound on garbage collection.
+         */
+        auto partitions = co_await get_partitions(as);
+        if (!partitions.has_value()) {
+            co_return std::unexpected(partitions.error());
+        }
+        if (partitions.value().partitions.empty()) {
+            co_return std::nullopt;
+        }
+
         /*
-         * TODO(noah): missing an integration of epochs with the partition
-         * table. for now we report no eligible epoch.
+         * Next we retrieve the latest reported epoch bounds from all cloud
+         * topic partitions. The source for this information is distributed,
+         * while the source for the `partitions` set above is centralized, and
+         * this is why we have these two different collection steps.
          */
-        co_return std::nullopt;
+        auto gc_epochs = co_await get_partitions_max_gc_epoch(as);
+        if (!gc_epochs.has_value()) {
+            co_return std::unexpected(gc_epochs.error());
+        }
+
+        /*
+         * The final result begins as the maximum epoch for the given snapshot.
+         * Below we merge the two result sets and walk the final result back to
+         * account for the partition with the smallest eligible gc epoch.
+         */
+        auto result = cluster_epoch(partitions.value().last_applied());
+
+        for (const auto& partition : partitions.value().partitions) {
+            const auto& tp_ns = partition.first;
+            auto nit = gc_epochs.value().find(tp_ns);
+            if (nit == gc_epochs.value().end()) {
+                co_return std::unexpected(
+                  fmt::format(
+                    "Topic '{}' in snapshot has no reported max GC epoch",
+                    tp_ns));
+            }
+
+            for (const auto p_id : partition.second) {
+                auto pit = nit->second.find(p_id);
+                if (pit == nit->second.end()) {
+                    co_return std::unexpected(
+                      fmt::format(
+                        "Partition '{}/{}' in snapshot has no reported max GC "
+                        "epoch",
+                        tp_ns,
+                        p_id));
+                }
+
+                // this partition may hold back the max GC eligible epoch
+                result = std::min(result, pit->second);
```

Comment:
I'm kind of surprised we don't have to do any validation of the revision_ids of the partitions. I can't immediately think of a case where this is problematic, but it does seem like non-obvious behavior that the values returned may not have anything to do with the snapshot we took

Reply:
The closest "weird" behavior I can think of is a scenario where the health report has some later topic revision than the one in the snapshot (e.g. because we are a stale controller leader). In this scenario, if the health reports were actually referring to a later revision of the topic, and so we may apply an epoch that doesn't make any sense for the partition from the controller. It's technically safe to do so because this older partition _should_ be GCed, but that mismatch seemed worth calling out.

I don't think there is a correctness concern, but it's non-obvious to me because we're operating on pieces of state that are snapshotted non-atomically with respect to one another, and we don't address in comments or otherwise why that's safe.

Reply:
> where the health report has some later topic revision than the one in the snapshot

to be clear you are referring to weird behavior where two topics with the same name are being compared but they are really different topic instances?

Reply:
I think we need to check if the snapshot that we created by looking at topic table is actually consistent. The controller offset of the snapshot is the upper bound for the GC epoch. Do we enforce this invariant somewhere?